### PR TITLE
 Deprecate PodTemplate and ServiceType in favour of Resource 

### LINKED ISSUE #883: Can't remove old EventListener in latest nightly
# Expected Behavior

`kubectl delete el X` should remove eventlistenter.
 
# Actual Behavior

Webhook fails to parse:

```
Error from server (BadRequest): error when deleting "tekton/trigger-bot.yaml": admission webhook "validation.webhook.triggers.tekton.dev" denied the request: decoding request failed: cannot decode incoming old object: json: unknown field "podTemplate"
```

# Steps to Reproduce the Problem

1. Create el with `podTemplate` on v0.10
2. Update to nightly
3. Delete/apply el

# Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

  ```
Client Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.0-rc.2", GitCommit:"27bb2a4a0a5cb8330178d19e57fa61fffa895c98", GitTreeState:"clean", BuildDate:"2020-07-21T17:39:35Z", GoVersion:"go1.14.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:51:19Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}  ```

- Tekton Pipeline version:

```
Client version: 0.15.0
Pipeline version: v0.19.0
Triggers version: v20201230-f9f9ad66e5
Dashboard version: v0.12.0
```


<!-- Any other additional information -->

### PR DESCRIPTION
**Pull request by khrm**
# Changes
Fixes #883

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#tests) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#docs) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commits)
- [ ] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
Deprecate PodTemplate and ServiceType in favour of Resource .
```

### CODE REVIEW

File: pkg/apis/triggers/v1alpha1/event_listener_defaults.go [hunk 1]
```diff
@@ -31,5 +31,37 @@ func (el *EventListener) SetDefaults(ctx context.Context) {
 			triggerSpecBindingArray(el.Spec.Triggers[i].Bindings).
 				defaultBindings()
 		}
+		// Remove Deprecated Resource Fields
+		el.Spec.updatePodTemplate()
+		el.Spec.updateServiceType()
+	}
+}
+
+// To be Removed in a later release
```
Comment (dibyom) [inline pkg/apis/triggers/v1alpha1/event_listener_defaults.go:40]: One small thing: Could we open an issue to track the removal?

File: test/builder/eventlistener.go [hunk 2]
```diff
@@ -103,6 +106,38 @@ func EventListenerReplicas(replicas int32) EventListenerSpecOp {
 	}
 }
 
+// EventListenerPodTemplate sets the specified pod template of the EventListener.
+func EventListenerPodTemplate(podTemplate v1alpha1.PodTemplate) EventListenerSpecOp {
+	return func(spec *v1alpha1.EventListenerSpec) {
+		spec.DeprecatedPodTemplate = podTemplate
+	}
```
Comment (dibyom) [inline test/builder/eventlistener.go:113]: are there any tests using this? If not, do we even need these given that the fields are deprecated?

File: pkg/apis/triggers/v1alpha1/event_listener_defaults.go [hunk 3]
```diff
@@ -31,5 +31,37 @@ func (el *EventListener) SetDefaults(ctx context.Context) {
...
+func (spec *EventListenerSpec) updatePodTemplate() {
```
Comment (dibyom) [inline pkg/apis/triggers/v1alpha1/event_listener_defaults.go:41]: could we add a couple of test cases for these in https://github.com/tektoncd/triggers/blob/master/pkg/apis/triggers/v1alpha1/event_listener_defaults_test.go#L28

**Review by dibyom — COMMENTED** [review]
Looking good. Just a couple of minor comments.

File: pkg/apis/triggers/v1alpha1/event_listener_defaults.go [hunk 4]
```diff
@@ -31,5 +31,37 @@ func (el *EventListener) SetDefaults(ctx context.Context) {
...
+	if spec.DeprecatedPodTemplate.NodeSelector != nil {
+		if spec.Resources.KubernetesResource == nil {
+			spec.Resources.KubernetesResource = &KubernetesResource{}
+		}
+		spec.Resources.KubernetesResource.WithPodSpec.Template.Spec.NodeSelector = spec.DeprecatedPodTemplate.NodeSelector
```
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/event_listener_defaults.go:46]:
```suggestion
		spec.Resources.KubernetesResource.Template.Spec.NodeSelector = spec.DeprecatedPodTemplate.NodeSelector
```

File: pkg/apis/triggers/v1alpha1/event_listener_defaults.go [hunk 5]
```diff
@@ -31,5 +31,37 @@ func (el *EventListener) SetDefaults(ctx context.Context) {
...
+		spec.DeprecatedPodTemplate.NodeSelector = nil
+	}
+	if spec.DeprecatedPodTemplate.Tolerations != nil {
+		if spec.Resources.KubernetesResource == nil {
+			spec.Resources.KubernetesResource = &KubernetesResource{}
+		}
+		spec.Resources.KubernetesResource.WithPodSpec.Template.Spec.Tolerations = spec.DeprecatedPodTemplate.Tolerations
```
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/event_listener_defaults.go:53]:
```suggestion
		spec.Resources.KubernetesResource.Template.Spec.Tolerations = spec.DeprecatedPodTemplate.Tolerations
```

File: pkg/apis/triggers/v1alpha1/event_listener_types.go [hunk 6]
```diff
@@ -53,11 +53,13 @@ type EventListener struct {
...
 // by a list of Triggers.
 type EventListenerSpec struct {
-	ServiceAccountName string                 `json:"serviceAccountName"`
-	Triggers           []EventListenerTrigger `json:"triggers"`
-	Replicas           *int32                 `json:"replicas,omitempty"`
-	NamespaceSelector  NamespaceSelector      `json:"namespaceSelector,omitempty"`
-	Resources          Resources              `json:"resources,omitempty"`
+	ServiceAccountName    string                 `json:"serviceAccountName"`
+	Triggers              []EventListenerTrigger `json:"triggers"`
+	DeprecatedServiceType corev1.ServiceType     `json:"serviceType,omitempty"`
```
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/event_listener_types.go:58]: as mentioned here https://github.com/tektoncd/triggers/pull/897/files?file-filters%5B%5D=.go#r556806397 can we add comment above `DeprecatedServiceType` and mention the created issue

File: pkg/apis/triggers/v1alpha1/event_listener_types.go [hunk 7]
```diff
@@ -53,11 +53,13 @@ type EventListener struct {
...
+	Replicas              *int32                 `json:"replicas,omitempty"`
+	DeprecatedPodTemplate PodTemplate            `json:"podTemplate,omitempty"`
```
Comment (savitaashture) [inline pkg/apis/triggers/v1alpha1/event_listener_types.go:60]: here as well

**Review by savitaashture — COMMENTED** [review]
looks good to me :+1: 

Few minor comments

**Pull request merged**